src/sleep/sleep.c

```diff
@@ -288,11 +288,19 @@ static int execute_s2h(const SleepConfig *sleep_config) {
                 else
                         return log_error_errno(r, "Error fetching battery capacity percentage: %m");
 
-                last_capacity = r;
-                before_timestamp = now(CLOCK_BOOTTIME);
-
-
-                log_debug("Set timerfd wake alarm to estimate battery discharge rate for %s", FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
+                r = get_battery_discharge_rate();
+                if ((last_capacity * 2) <= r)
+                        return 0;
```

Comment:
it needs to break out of the loop here to hibernate, returning means it will just resume